[OpenMP] Introduce -fopenmp-force-usm flag

jplehr opened the pull request:
The new flag implements logic to include `#pragma omp requires unified_shared_memory` in every translation unit.
This enables a straightforward way to enable USM for an application without the need to modify sources.

This is the flag mentioned in https://github.com/llvm/llvm-project/pull/75467
Once the test landed, I'll rebase and enable the test with this patch.

jplehr committed:
[OpenMP] Introduce -fopenmp-force-usm flag

The new flag implements logic to include #pragma omp requires
unified_shared_memory in every translation unit.
This enables a straightforward way to enable USM for an application
without the need to modify sources.

jhuber6 commented on clang/lib/Driver/ToolChains/AMDGPUOpenMP.cpp:
```diff
@@ -129,6 +129,22 @@ AMDGPUOpenMPToolChain::GetCXXStdlibType(const ArgList &Args) const {
+  // unified_shared_memory via the force_usm header
+  // XXX This may result in a compilation error if the source
+  // file already includes that pragma.
+  if (DriverArgs.hasArg(options::OPT_fopenmp_force_usm)) {
+    CC1Args.push_back("-include");
+    CC1Args.push_back(
+        DriverArgs.MakeArgString(HostTC.getDriver().ResourceDir +
+                                 "/include/openmp_wrappers/force_usm.h"));
```
I don't think this is a good way to handle this. We should make this a CC1 argument, forward it in the standard way, and make `CGOpenMPRuntime` always emit the associated runtime call.

Also note that I'm planning on removing the current "requires" handling because emitting spurious global constructors into the runtime is difficult to work around.

jhuber6 replied:
Here's the patch for `-fopenmp-offload-mandatory` which is a similar use-case https://reviews.llvm.org/D120353.

jhuber6 commented on clang/include/clang/Driver/Options.td:
```diff
@@ -3381,6 +3381,8 @@ def fopenmp_cuda_blocks_per_sm_EQ : Joined<["-"], "fopenmp-cuda-blocks-per-sm=">
   Flags<[NoArgumentUnused, HelpHidden]>, Visibility<[ClangOption, CC1Option]>;
 def fopenmp_cuda_teams_reduction_recs_num_EQ : Joined<["-"], "fopenmp-cuda-teams-reduction-recs-num=">, Group<f_Group>,
   Flags<[NoArgumentUnused, HelpHidden]>, Visibility<[ClangOption, CC1Option]>;
+def fopenmp_force_usm : Flag<["-"], "fopenmp-force-usm">, Group<f_Group>,
+  Flags<[NoArgumentUnused, HelpHidden]>, Visibility<[CC1Option]>;
```
`-f` options tend to have a `-fno` variant as well.

jplehr replied:
With the intent to remove the USM behavior from a codebase that has the requires pragma, by basically just ignoring it?

jplehr committed:
[OpenMP] Implicit USM Clause Solution

This uses an implicitly added OpenMP USM Clause when initializing SEMA
to enforce the use of USM.

jplehr committed:
Revert "[OpenMP] Introduce -fopenmp-force-usm flag"

This reverts commit 4ecd07d786a5a994b33b9177d4e21d839bfe3fc9.

To test the other solution.

jplehr commented:
Hmm.. I guess I screwed something up with git and the history.
